File: mobile/src/debug/java/jp/satorufujiwara/kotlin/DebugAppModule.java
package jp.satorufujiwara.kotlin;

import com.squareup.leakcanary.LeakCanary;
import com.squareup.leakcanary.RefWatcher;

import android.app.Application;
import android.support.annotation.NonNull;

import dagger.Module;
import dagger.Provides;

@Module
public class DebugAppModule {

    private final DaggerApp app;

    public DebugAppModule(@NonNull final DaggerApp app) {
        this.app = app;
    }

    @Provides
    @AppScope
    Application provideApplication() {
        return app;
    }

    @Provides
    @AppScope
    public AppLifecycleCallbacks provideAppLifecycleCallbacks() {
        return new DebugAppLifecycleCallbacks();
    }

    @Provides
    public RefWatcher provideRefWatcher(Application application) {
        return LeakCanary.install(application);
    }
}

File: mobile/src/debug/kotlin/jp/satorufujiwara/kotlin/DebugAppModule.kt
package jp.satorufujiwara.kotlin

import android.app.Application
import com.squareup.leakcanary.LeakCanary
import com.squareup.leakcanary.RefWatcher
import dagger.Module
import dagger.Provides

@Module
class DebugAppModule(private val app: Application) {

    @Provides
    @AppScope
    fun provideApplication(): Application = app

    @Provides
    @AppScope
    fun provideAppLifecycleCallbacks(): AppLifecycleCallbacks = DebugAppLifecycleCallbacks()

    @Provides
    @AppScope
    fun provideRefWatcher(application: Application): RefWatcher =
            LeakCanary.install(application)
}
